File: backend/services/stigmer-server/pkg/domain/workflow/validation/crossref_test.go
package validation

import (
	"strings"
	"testing"

	workflowv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflow/v1"
	"google.golang.org/protobuf/types/known/structpb"
)

func makeTask(name string, kind workflowv1.WorkflowTaskKind, config map[string]interface{}) *workflowv1.WorkflowTask {
	cfg, _ := structpb.NewStruct(config)
	return &workflowv1.WorkflowTask{
		Name:       name,
		Kind:       kind,
		TaskConfig: cfg,
	}
}

func TestValidateCrossTaskReferences_UniqueNames(t *testing.T) {
	cfg := map[string]interface{}{"variables": map[string]interface{}{"x": "1"}}
	spec := &workflowv1.WorkflowSpec{
		Tasks: []*workflowv1.WorkflowTask{
			makeTask("step_a", workflowv1.WorkflowTaskKind_set_vars, cfg),
			makeTask("step_a", workflowv1.WorkflowTaskKind_set_vars, cfg),
		},
	}

	errors := ValidateCrossTaskReferences(spec)
	if len(errors) == 0 {
		t.Error("Expected duplicate name error")
	}
	if !strings.Contains(errors[0], "duplicate task name") {
		t.Errorf("Expected 'duplicate task name' error, got: %s", errors[0])
	}
}

func TestValidateCrossTaskReferences_InvalidFlowThen(t *testing.T) {
	cfg := map[string]interface{}{"variables": map[string]interface{}{"x": "1"}}
	spec := &workflowv1.WorkflowSpec{
		Tasks: []*workflowv1.WorkflowTask{
			{
				Name:       "step_a",
				Kind:       workflowv1.WorkflowTaskKind_set_vars,
				TaskConfig: mustStruct(cfg),
				Flow:       &workflowv1.FlowControl{Then: "nonexistent"},
			},
		},
	}

	errors := ValidateCrossTaskReferences(spec)
	if len(errors) == 0 {
		t.Error("Expected unknown task reference error")
	}
	if !strings.Contains(errors[0], "unknown task 'nonexistent'") {
		t.Errorf("Expected unknown task error, got: %s", errors[0])
	}
}

func TestValidateCrossTaskReferences_ValidFlowEnd(t *testing.T) {
	cfg := map[string]interface{}{"variables": map[string]interface{}{"x": "1"}}
	spec := &workflowv1.WorkflowSpec{
		Tasks: []*workflowv1.WorkflowTask{
			{
				Name:       "step_a",
				Kind:       workflowv1.WorkflowTaskKind_set_vars,
				TaskConfig: mustStruct(cfg),
				Flow:       &workflowv1.FlowControl{Then: "end"},
			},
		},
	}

	errors := ValidateCrossTaskReferences(spec)
	if len(errors) > 0 {
		t.Errorf("Expected no errors for flow.then='end', got: %v", errors)
	}
}

func TestValidateCrossTaskReferences_CycleDetection(t *testing.T) {
	cfg := map[string]interface{}{"variables": map[string]interface{}{"x": "1"}}
	spec := &workflowv1.WorkflowSpec{
		Tasks: []*workflowv1.WorkflowTask{
			{
				Name:       "step_a",
				Kind:       workflowv1.WorkflowTaskKind_set_vars,
				TaskConfig: mustStruct(cfg),
				Flow:       &workflowv1.FlowControl{Then: "step_b"},
			},
			{
				Name:       "step_b",
				Kind:       workflowv1.WorkflowTaskKind_set_vars,
				TaskConfig: mustStruct(cfg),
				Flow:       &workflowv1.FlowControl{Then: "step_a"},
			},
		},
	}

	errors := ValidateCrossTaskReferences(spec)
	hasCycleError := false
	for _, e := range errors {
		if strings.Contains(e, "circular dependency") {
			hasCycleError = true
			break
		}
	}
	if !hasCycleError {
		t.Errorf("Expected cycle detection error, got: %v", errors)
	}
}

func TestValidateCrossTaskReferences_ValidWorkflow(t *testing.T) {
	cfg := map[string]interface{}{"variables": map[string]interface{}{"x": "1"}}
	spec := &workflowv1.WorkflowSpec{
		Tasks: []*workflowv1.WorkflowTask{
			{
				Name:       "step_a",
				Kind:       workflowv1.WorkflowTaskKind_set_vars,
				TaskConfig: mustStruct(cfg),
				Flow:       &workflowv1.FlowControl{Then: "step_b"},
			},
			{
				Name:       "step_b",
				Kind:       workflowv1.WorkflowTaskKind_set_vars,
				TaskConfig: mustStruct(cfg),
				Flow:       &workflowv1.FlowControl{Then: "end"},
			},
		},
	}

	errors := ValidateCrossTaskReferences(spec)
	if len(errors) > 0 {
		t.Errorf("Expected no errors for valid workflow, got: %v", errors)
	}
}

func TestValidateCrossTaskReferences_DidYouMean(t *testing.T) {
	cfg := map[string]interface{}{"variables": map[string]interface{}{"x": "1"}}
	spec := &workflowv1.WorkflowSpec{
		Tasks: []*workflowv1.WorkflowTask{
			{
				Name:       "step_a",
				Kind:       workflowv1.WorkflowTaskKind_set_vars,
				TaskConfig: mustStruct(cfg),
				Flow:       &workflowv1.FlowControl{Then: "step_ab"},
			},
			{
				Name:       "step_b",
				Kind:       workflowv1.WorkflowTaskKind_set_vars,
				TaskConfig: mustStruct(cfg),
			},
		},
	}

	errors := ValidateCrossTaskReferences(spec)
	if len(errors) == 0 {
		t.Error("Expected error for typo reference")
	}
	hasSuggestion := false
	for _, e := range errors {
		if strings.Contains(e, "did you mean") {
			hasSuggestion = true
			break
		}
	}
	if !hasSuggestion {
		t.Errorf("Expected 'did you mean' suggestion, got: %v", errors)
	}
}

func mustStruct(m map[string]interface{}) *structpb.Struct {
	s, _ := structpb.NewStruct(m)
	return s
}

// The workflow surface accepts git_repo workspace sources only (no client
// is connected to serve a local_path when a task fires) — the one bespoke
// task-config rule left after stigmer#805 armed the declared proto rules.
// Source presence and git_repo.url's HTTPS shape are proto facts now enforced
// by ValidateTaskConfigConstraints (probes in task_config_constraints_test.go).
// Error strings are pinned in lockstep with the cloud Java validator.
func TestValidateTaskConfigSurfaceRules_AgentCallWorkspaceEntries(t *testing.T) {
	makeSpec := func(entries []interface{}) *workflowv1.WorkflowSpec {
		return &workflowv1.WorkflowSpec{
			Tasks: []*workflowv1.WorkflowTask{
				makeTask("review", workflowv1.WorkflowTaskKind_agent_call, map[string]interface{}{
					"agent":             "test-agent",
					"message":           "test message",
					"workspace_entries": entries,
				}),
			},
		}
	}

	t.Run("git https entry passes", func(t *testing.T) {
		errors := ValidateTaskConfigSurfaceRules(makeSpec([]interface{}{
			map[string]interface{}{
				"name": "app",
				"source": map[string]interface{}{
					"git_repo": map[string]interface{}{"url": "https://github.com/acme/app"},
				},
			},
		}))
		if len(errors) != 0 {
			t.Errorf("expected no errors, got %v", errors)
		}
	})

	t.Run("local_path source is rejected", func(t *testing.T) {
		errors := ValidateTaskConfigSurfaceRules(makeSpec([]interface{}{
			map[string]interface{}{
				"source": map[string]interface{}{
					"local_path": map[string]interface{}{"path": "/home/me/repo"},
				},
			},
		}))
		want := "task 'review' (agent_call): workspace_entries[0] must use a git_repo source — no client is connected to serve a local_path when a workflow task fires"
		if len(errors) != 1 || errors[0] != want {
			t.Errorf("expected [%q], got %v", want, errors)
		}
	})

	t.Run("missing source is the constraints step's report, not this one's", func(t *testing.T) {
		errors := ValidateTaskConfigSurfaceRules(makeSpec([]interface{}{
			map[string]interface{}{"name": "app"},
		}))
		if len(errors) != 0 {
			t.Errorf("expected no errors (WorkspaceEntry.source's required rule reports absence), got %v", errors)
		}
	})

	t.Run("second entry's index is reported", func(t *testing.T) {
		errors := ValidateTaskConfigSurfaceRules(makeSpec([]interface{}{
			map[string]interface{}{
				"source": map[string]interface{}{
					"git_repo": map[string]interface{}{"url": "https://github.com/acme/app"},
				},
			},
			map[string]interface{}{
				"source": map[string]interface{}{
					"local_path": map[string]interface{}{"path": "/tmp/x"},
				},
			},
		}))
		if len(errors) != 1 || !strings.Contains(errors[0], "workspace_entries[1]") {
			t.Errorf("expected one error naming workspace_entries[1], got %v", errors)
		}
	})
}

File: backend/services/stigmer-server/src/domain/workflow/validation/__tests__/crossref.test.ts
/**
 * Cross-reference validation tests — pin the Go crossref.go behavior:
 * unknown kinds, duplicate names, flow.then / fallback_task /
 * cases[].then / outcomes[].then resolution with did-you-mean, "end"
 * sentinel handling, cycle detection, and the workspace-entries surface
 * rule. Message strings are the cross-edition contract.
 */
import { create } from "@bufbuild/protobuf";
import type { JsonObject } from "@bufbuild/protobuf";
import { describe, expect, it } from "vitest";

import { WorkflowTaskKind } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/enum_pb";
import { WorkflowSpecSchema } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/spec_pb";
import type { WorkflowSpec } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/spec_pb";

import {
  validateCrossTaskReferences,
  validateTaskConfigSurfaceRules,
  validateTaskKinds,
} from "../crossref.js";

function spec(
  tasks: Array<{
    name: string;
    kind: WorkflowTaskKind;
    taskConfig?: JsonObject;
    flow?: { then: string };
  }>,
): WorkflowSpec {
  return create(WorkflowSpecSchema, {
    document: { dsl: "1.0.0", namespace: "t", name: "t", version: "0.1.0" },
    tasks,
  });
}

const SET = (name: string, then?: string) => ({
  name,
  kind: WorkflowTaskKind.set_vars,
  taskConfig: { variables: { k: "v" } } as JsonObject,
  ...(then !== undefined ? { flow: { then } } : {}),
});

describe("validateTaskKinds", () => {
  it("rejects unspecified and unknown kind values with the pinned message", () => {
    const s = spec([
      { name: "z", kind: 0 as WorkflowTaskKind },
      { name: "u", kind: 9999 as WorkflowTaskKind },
    ]);
    expect(validateTaskKinds(s)).toEqual([
      "task 'z': unknown or unspecified task kind (value=0)",
      "task 'u': unknown or unspecified task kind (value=9999)",
    ]);
  });

  it("accepts every known kind", () => {
    expect(validateTaskKinds(spec([SET("a")]))).toEqual([]);
  });
});

describe("validateCrossTaskReferences", () => {
  it("reports duplicate task names with both indices", () => {
    const errors = validateCrossTaskReferences(spec([SET("dup"), SET("dup")]));
    expect(errors).toContain(
      'duplicate task name "dup" at tasks[1]: already defined at tasks[0]',
    );
  });

  it("resolves flow.then, honors the end sentinel, suggests near-misses", () => {
    const errors = validateCrossTaskReferences(
      spec([SET("start", "finsih"), SET("finish", "end")]),
    );
    expect(errors).toEqual([
      "task 'start' flow.then references unknown task 'finsih' (did you mean 'finish'?)",
    ]);
  });

  it("omits the suggestion when nothing is within distance 3", () => {
    const errors = validateCrossTaskReferences(
      spec([SET("start", "wholly-unrelated-target")]),
    );
    expect(errors).toEqual([
      "task 'start' flow.then references unknown task 'wholly-unrelated-target'",
    ]);
  });

  it("checks llm_call/validate fallback_task and agent_call output.fallback_task", () => {
    const errors = validateCrossTaskReferences(
      spec([
        {
          name: "llm",
          kind: WorkflowTaskKind.llm_call,
          taskConfig: { model: "m", prompt: "p", fallback_task: "nope" },
        },
        {
          name: "val",
          kind: WorkflowTaskKind.validate,
          taskConfig: { input: "${ . }", fallbackTask: "nope" },
        },
        {
          name: "agent",
          kind: WorkflowTaskKind.agent_call,
          taskConfig: {
            agent: "a",
            message: "m",
            output: { fallback_task: "nope" },
          },
        },
      ]),
    );
    expect(errors).toEqual([
      "task 'llm' (llm_call) fallback_task references unknown task 'nope'",
      "task 'val' (validate) fallback_task references unknown task 'nope'",
      "task 'agent' (agent_call) output.fallback_task references unknown task 'nope'",
    ]);
  });

  it("checks switch cases[].then and human_input outcomes[].then with end allowed", () => {
    const errors = validateCrossTaskReferences(
      spec([
        {
          name: "sw",
          kind: WorkflowTaskKind.switch_case,
          taskConfig: {
            cases: [{ then: "missing" }, { then: "end" }],
          },
        },
        {
          name: "gate",
          kind: WorkflowTaskKind.human_input,
          taskConfig: {
            prompt: "p",
            outcomes: [{ name: "ok", then: "alsoMissing" }, { name: "no", then: "end" }],
          },
        },
      ]),
    );
    expect(errors).toEqual([
      "task 'sw' (switch_case) cases[0].then references unknown task 'missing'",
      "task 'gate' (human_input) outcomes[0].then references unknown task 'alsoMissing'",
    ]);
  });

  it("detects flow.then cycles", () => {
    const errors = validateCrossTaskReferences(
      spec([SET("a", "b"), SET("b", "a")]),
    );
    expect(errors).toHaveLength(1);
    expect(errors[0]).toMatch(/^circular dependency detected: (a -> b -> a|b -> a -> b)$/);
  });

  it("accepts an acyclic graph terminating in end", () => {
    expect(
      validateCrossTaskReferences(spec([SET("a", "b"), SET("b", "end")])),
    ).toEqual([]);
  });
});

describe("validateTaskConfigSurfaceRules", () => {
  it("refuses non-git_repo workspace sources on agent_call, skips absent sources", () => {
    const errors = validateTaskConfigSurfaceRules(
      spec([
        {
          name: "call",
          kind: WorkflowTaskKind.agent_call,
          taskConfig: {
            agent: "a",
            message: "m",
            workspace_entries: [
              { name: "ok", source: { git_repo: { url: "https://x.test/r" } } },
              { name: "local", source: { local_path: { path: "/tmp" } } },
              { name: "absent" },
            ],
          },
        },
      ]),
    );
    expect(errors).toEqual([
      "task 'call' (agent_call): workspace_entries[1] must use a git_repo source — no client is connected to serve a local_path when a workflow task fires",
    ]);
  });
});
